### DISCUSSION TITLE
Learning from "competing" formats

### DISCUSSION THREAD

**Discussion by MikeRalphson** [OAI/sig-moonwalk#47] [General]
* [fern](https://docs.buildwithfern.com/comparison-with-openapi) (new to me this day)
* [Amazon smithy](https://smithy.io/2.0/index.html) more documentation and code seems available for Smithy 2.0 e.g. [converting to OAS](https://smithy.io/2.0/guides/converting-to-openapi.html#smithy-to-openapi)
* RAML - traits and overlays already under consideration, more?

**Comment by darrelmiller** [comment]
I'll throw this https://github.com/microsoft/typespec into the mix too (formerly known as cadl).

Also, I think Steven Mizell book has some great insight here https://smizell.com/language-oriented-approach/

**Reply by rafalkrupinski** [reply]
Not a big fan of handling importing of javascript files into my python program

https://typespec.io/docs/language-basics/imports/

**Reply by SensibleWood** [reply]
Imports:

* Import TypeSpec: Another TypeSpec file you've built your model on.
* Import JavaScript: Some way you've extended TypeSpec, which you'll use in your model.

In essence, nothing ends up in your Python program - you are only importing to your TypeSpec model. You'd use an emitter to generate Python from the model that you "could" use in your implementation, if you chose to work that way.

**Reply by rafalkrupinski** [reply]
If my code generator was to use such typespec, it would need to know what functions this import brings and what those functions do

**Comment by JoanPuig** [comment]
I would suggest taking a look at: https://github.com/FIXTradingCommunity/fix-orchestra-spec, A lot of the design discussions around scenarios, parameter validation and workflows are very similar.

**Comment by handrews** [comment]
[JSON TypeSchema](https://typeschema.org/) (which I have not seen before) was brought up in an old spec issue as an alternative schema candidate:

* https://github.com/OAI/OpenAPI-Specification/issues/2839

**Reply by earth2marsh** [reply]
Lexicon and its Data Model specification is another example to consider: https://atproto.com/guides/lexicon

**Reply by rafalkrupinski** [reply]
Seems TypeSchema lacks any means of describing union types or even optional properties.

**Reply by SensibleWood** [reply]
> Seems TypeSchema lacks any means of describing union types or even optional properties.

```
Something? Something;
```

☝️ The question mark denotes optional. Without, mandatory

Unions: https://typespec.io/docs/language-basics/unions/

**Reply by rafalkrupinski** [reply]
@SensibleWood typespec vs typeschema

**Reply by SensibleWood** [reply]
Ah my mistake. Apologies

**Reply by rafalkrupinski** [reply]
easy mistake

**Comment by handrews** [comment]
CDDL is a data definition option, applying to either JSON or CBOR (CBOR support would be a nice bonus).  It is not in JSON form, although we could either import them as separate documents or create a JSON representation if we really wanted to.  CDDL 2 includes validation, annotation, and transformation on its road map:

* https://www.rfc-editor.org/rfc/rfc8610.html
* https://www.ietf.org/archive/id/draft-bormann-cbor-cddl-2-draft-04.html#name-processing-model-beyond-val

**Reply by earth2marsh** [reply]
See also: https://github.com/OAI/OpenAPI-Specification/issues/3108
